Simulator/simulator.py

```diff
@@ -95,324 +85,687 @@ def __init__(self, etype: etype, time: datetime, job_id: int):
+            # Add back the capacity
+            _, _, _, node_ids = self.current_reserve_jobs.pop(0)
+            for node_id in node_ids:
+                self._reserve_nodes[node_id].remove_job(next_dep.job_id)
         else:
-          # If Job is long, and we cannot schedule, then wait
-          self.wait_queue.append(cur_job.job_id)
-      else:
-        # A job is finished. So, check the wait queue
-        while len(self.wait_queue) > 0:
-          job_id = self.wait_queue[0]
-          waiting_job = self.input_trace.iloc[job_id]
-          time_waited_sec = (
-                  next_event.time - waiting_job.time
-          ).total_seconds()
-
-          # Compute maximum balking time
-          max_wait_time_sec = max_wait_time_min * 60
-
-          # Check if job waited more than the balking time
-          if time_waited_sec > max_wait_time_sec:
-            # Remove the job from the queue
-            self.wait_queue.popleft()
-            self.wait_time_map[job_id] = max_wait_time_sec
-            self.job_schedule_map[job_id] = "D"
-
-            end = max(
-              end, next_event.time + timedelta(seconds=int(runtime_sec))
-            )
-          else:
-            # Try scheduling the job
-            cpu_req = waiting_job.num_cores
-            runtime_sec = waiting_job.wallclock_runtime_sec
-            if self._schedule_job(
-                    next_event.time, cpu_req, runtime_sec, job_id
-            ):
-              # remove the job from the queue
-              self.wait_queue.popleft()
-
-              # Add waiting time to wait times map
-              self.wait_time_map[job_id] = time_waited_sec
-
-              end = max(
-                end,
-                next_event.time + timedelta(seconds=int(runtime_sec)),
-                )
+            next_event = Event(etype.schedule, next_arr_time, next_arr.job_id)
+            self.next_jobID += 1
+
+        return next_event
+
+    def is_short_wait(self, model, cur_job) -> bool:
+        X_test = self.calc_system_state_features(cur_job)
+        return model.predict(X_test)
+
+    def queue_is_empty(self) -> bool:
+        return not self.wait_queue
+
+    def set_waiting_time(self, job, waiting_time):
+        self.wait_time_map[job.job_id] = waiting_time
+
+    def calc_waiting_time_and_reserved_cost(self):
+        self.input_trace["wait_time_sec"] = (self.input_trace["job_id"].map(self.wait_time_map))
+        self.mean_waiting_time, self.reserved_cost = self.input_trace.wait_time_sec.mean(), self._compute_fixed_cost()
+
+    """
+    Calculates approximate cost of running fixed cluster for trace duration
+    """
+    def _compute_fixed_cost(self) -> float:
+        self.input_trace['submit_time'] = pd.to_numeric(self.input_trace['submit_time'])
+        self.input_trace['end_time'] = pd.to_numeric(self.input_trace['end_time'])
+
+        start_time = self.input_trace.loc[self.input_trace['submit_time'].idxmin()].submit_time
```

Comment:
is the idxmin + idxmax clause enough? Do we need to run an additional check to see if its run in on demand resource?

Reply:
yeah this should be fine. the thought is that you'll have the reserved resources for the duration of the trace regardless of how many jobs are actually run on fixed resources since it's reserved capacity. technically the duration of the trace might be a little different (like a couple minutes) depending on the scheduling policies you use, but we're not actually using the fixed cost figure anywhere in the report (just included it for our reference), so this should be good enough